pg_search/src/postgres/storage/fsm.rs

```diff
@@ -34,413 +35,1263 @@ enum FSMBlockKind {
+/// This value is the block number that represents the start of the list of block numbers that are
+/// free for the xid key stored in that slot.  This value stays with the slot, even as the tree is
+/// mutated as generally speaking, with an array-backed AVL tree entries don't move slots, only their
+/// left/right pointers are updated during rebalancing.  Special consideration is made for deleting
+/// an entry, however, when the key/value is moved to a new slot -- in this case the tag values are
+/// swapped until the entry is finished moving.  Essentially, tags have an affinity for their key/value
+/// as long as the slot is occupied.
+///
+/// The FSM has two user-facing operations: extend and drain.
+///
+/// # Extend
+///
+/// The caller provides both a transaction id and an iterator of free blocks to associate with that
+/// transaction.  That transaction indicates the point in time in which those blocks are usable.  Any
+/// future transaction `>=` to that transaction will be able to use those blocks.
+///
+/// During an extension a share lock is acquired on the FSM's root page (the AVL tree) and the
+/// specified transaction id is searched.  If it's found, then an exclusive lock is taken on the block
+/// represented by that xid's `tag` -- the blocklist -- and the blocklist is extended from head-to-tail,
+/// filling in gaps along the way.  However, if a full page is encountered, then a shortcut is taken
+/// and a new blocklist is linked into the existing list at that point.
+///
+/// Concurrent extensions are allowed on the same xid, blocking one block in the block list at a time.
+/// However, in practice almost every call to [`FreeSpaceManager::extend_with_when_recyclable`] uses
+/// the current transaction id, and since each transaction has its own unique id, there'll be no blocking.
+///
+/// Sometimes a caller will use the result of [`pg_sys::ReadNextFullTransactionId()`] instead of the
+/// current transaction id.  This can cause concurrent extensions of the same transaction id.  This
+/// is fine and generally a rare situation.
+///
+/// If the provided transaction id is not found, then the shared lock on the FSM root is upgraded to
+/// an exclusive lock and the transaction id is inserted.  It is possible that by this time the
+/// transaction id now exists.  In either case, the `tag` value for the new (or now-existing) transaction
+/// id's slot is used as the blocklist, and it is extended.
+///
+/// The upgraded exclusive lock on the tree is *not* held during extension.  It only lives long enough
+/// to ensure proper insertion of the possibly new transaction id.
+///
+/// If, when extending a blocklist, another page is needed, it is acquired by extending the relation by
+/// a page, not by asking the FSM itself for a page.
+///
+/// ## What happens if the tree is full?
+///
+/// The AVL tree is backed by a fixed-size array of (at the time of writing) 338 slots.  If all the
```

Comment:
338 slots * 2000 blocks in a freelist * 8k per block == 5.4GB of FSM capacity, right? That's assuming too that the freelists are all full, which sounds unlikely.

Is merging incremental to the point where all of the blocks of a segment will never be in the FSM at once?